Batteries: add max charge and discharge power

Hofyyy opened the pull request:
- Added maxchargepower
- Added maxdischargepower
- Removed advanced: true if found

Hofyyy committed:
Added maxcharge and discharge power to all meter with capacity

- Added maxchargepower
- Added maxdischargepower
- Removed advanced: true if found

Hofyyy committed:
Merge branch 'master' into battery_charge_power

CiNcH83 commented:
Although admittedly I wasn't sure either, [this](https://github.com/evcc-io/evcc/pull/23555/files) seems to work fine for me with Huawei. Guess the assignment is still needed to pass the variable on to the internal logic.

andig commented:
> Removed advanced: true if found

Please don't. It's there for a reason.

Hofyyy commented:
> > Removed advanced: true if found
> 
> Please don't. It's there for a reason.

Kannst du so nett sein und folgende Datei reviewen im Bezug auf das advanced true.
Ich denke meine Änderung ist richtig, sonst verstehe ich nicht warum wir es woanders weglassen können.
templates/definition/meter/sonnenbatterie_eco56.yaml

Hofyyy committed:
Update huawei-sun2000.yaml

- Add saving of the maxchargepower value at the end, because it is only used for set the battery at the moment.

Hofyyy committed:
Update deye-storage.yaml

- Added save of charge and discharge power at the end. This was forgotten.

Hofyyy committed:
Update kostal-piko-hybrid.yaml

- Forgot saving of max charge and discharge ...

Hofyyy committed:
Update marstek-venus.yaml

- Save maxchargepower at the end, because it corrently only used to set the battery.

Hofyyy committed:
Update openems.yaml

- Saved maxchargepower at the end.

Hofyyy committed:
Update sma-hybrid.yaml

- Save at the end maxdischargepower

Hofyyy committed:
Update sma-sbs-15-25-modbus.yaml

saved: maxchargepower at the end

Hofyyy committed:
Update sma-sbs-modbus.yaml

Saved maxchargepower at the end.

Hofyyy committed:
Update sma-si-modbus.yaml

- Saved maxchargepower at the end.

Hofyyy committed:
Update sonnenbatterie.yaml

- Saved also maxchargepower at the end.

Hofyyy committed:
Update sonnenbatterie_eco56.yaml

- Saved also maxchargepower at the end.

Hofyyy committed:
Update sungrow-hybrid.yaml

- saved also maxchargepower at the end.

Hofyyy committed:
Update varta.yaml

- saved also maxdischargepower at the end.

Hofyyy commented:
> @andig sofern ich das richtig gemacht habe und du den powerwall change als sinnvoll ansiehst, kann ich gerne einen PR dafür eröffnen. Hier scheint die powerwall aktuell noch nicht mit drin zu sein.



> @Hofyyy ich habe maxChargePower und maxDischargePower für lokale Tests einmal prototypisch für die Tesla Powerwall hinzugefügt, eventuell hilft es zur Orientierung:
> 
> [mfuchs1984@3e62abc#diff-55842ed79502b62e399739bfdaa72767582f328158ba78170c1826727b7f39c4](https://github.com/mfuchs1984/evcc/commit/3e62abc292bf3ed21980440e8f3a340d1812777a#diff-55842ed79502b62e399739bfdaa72767582f328158ba78170c1826727b7f39c4)
> 
> Editieren musste ich powerwall.go und tesla-powerwall.yaml. Danach per `make assets` powerwall_decorators.go generieren.

Vielen Dank für den Tipp.
Ich muss auch erstmal gucken warum ich die Powerwall nicht in meinen Änderungen drin habe. Die Idee war ja gerade, dass ich einmal alle Batterien durchgehe ...

andig commented:
Sieht gut aus 👍🏻

Hofyyy committed:
Add LG ESS information

- Add LG ESS information

Hofyyy committed:
Removed comment

Removed comment, because the compiler does not like it.

Hofyyy committed:
fixed decorateLgEss

Hofyyy committed:
Test e3dc

Hofyyy committed:
fixed compile

mfuchs1984 commented:
> Ich habe einen test mit deinen Änderungen geradeziehen können. E3DC ist leider anders. 

Kannst du das genauer ausführen?

mfuchs1984 commented:
Lass mal im evcc Ordner `make lint` laufen.

Hofyyy commented:
> Lass mal im evcc Ordner `make lint` laufen.

meter/bosch_bpts5_hybrid.go:1: : # github.com/evcc-io/evcc/meter [github.com/evcc-io/evcc/meter.test]
meter/e3dc.go:75:112: cannot use cc.batteryPowerLimits.Decorator() (value of type func() (float64, float64)) as batteryPowerLimits value in argument to NewE3dc
meter/e3dc.go:116:25: cannot use batteryPowerLimits (variable of struct type batteryPowerLimits) as func() (float64, float64) value in assignment (typecheck)
package meter

mfuchs1984 committed:
Fix lint error in e3dc.go

Hofyyy committed:
Merge pull request #1 from mfuchs1984/patch-10

Fix lint error in e3dc.go

Hofyyy commented:
> Sollten nicht die anderen go-files auch noch angepasst werden?

Da hat die Überprüfung nicht angeschlagen, daher bin ich davon ausgegangen, das diese in Ordnung sind ...

Hofyyy commented:
> …und analog brauchts dann bei Bedarf auch noch einen PR für min/max soc. 

Ich bin froh wenn wir den erstmal durch die Tür bekommen :-). Aber ja.

mfuchs1984 commented:
Leg du gerne vor :)

Hofyyy committed:
Added bosch_bpts5_hyprid

Hofyyy committed:
Added goodwe-wifi

Hofyyy committed:
Add rct

Hofyyy committed:
added sma, fixed rct

Hofyyy commented:
Neue Runde, neues Glück. Bitte um Review. Ich bin einmal alle meter mit *_decoratoren durchgegangen und habe diese angepasst + linting. Des Weiteren habe ich mir jetzt gespart immer ins Datasheed zu gucken und diese bei 4600KW speed gelassen, da es sich von größe zu größe unterscheidet. Hier ist die Frage ob der default Speed relevant ist.

andig commented:
Am liebsten weglassen. Kein default== nix konfiguriert und damit auch keine Verhaltensänderung.

Hofyyy commented:
> Am liebsten weglassen. Kein default== nix konfiguriert und damit auch keine Verhaltensänderung.

So leid es mir tut, aber ich verstehe deinen Kommentar nicht. Ich habe nur die Dateien angefasst, die eine Kapazität haben und zu denen wir auch die yaml geändert haben. Kannst du mir bitte genauer sagen was zu viel des guten ist?

Hofyyy commented:
> Am liebsten weglassen. Kein default== nix konfiguriert und damit auch keine Verhaltensänderung.

Ok in den *.yaml Datei ist kein Default speed *.W. Aber in den go Dateien die wofür verwendet werden? Kann das da auch weg, oder braucht man das für Tests?

Hofyyy commented:
Gut aber kann ich die initialisierung einfach löschen, oder knallt es dann wenn man es nicht initialisiert?

andig commented on meter/lgess.go:
```diff
@@ -45,6 +46,10 @@ func NewLgEssFromConfig(other map[string]interface{}, essType lgpcs.Model) (api.
 			MaxSoc: 95,
 		},
+		batteryPowerLimits: batteryPowerLimits{
+			MaxChargePower:    5000,
```
Dito. Wenn das nicht 100% sicher ist lieber weglassen

Hofyyy committed:
Removed MaxCharge, MaxDischarge defaults

I removed the defaults in the go implementation, because I think they are better in the *.yaml file anyway.

hybe2022 commented on meter/sma.go:
```diff
@@ -24,16 +24,21 @@ func init() {
 		Scale                    float64 // power only
 	}{
+		batteryPowerLimits: batteryPowerLimits{
+			MaxChargePower:    4600,
```
Wo kommen diese niedrigen max Werte her. Ich hoffe dadurch wird nachher das Laden und Entladen nicht auf 0,1C bei mir reduziert. Ich würde schon gerne die maximalen 10,6kW meiner WR/Speichert Kombination nutzen...

Hofyyy committed:
Merge remote-tracking branch 'upstream/master' into battery_charge_power

andig commented on meter/e3dc.go:
```diff
@@ -28,19 +28,20 @@ func init() {
-		Key             string
-		DischargeLimit  uint32
-		Timeout         time.Duration
+		batteryCapacity    `mapstructure:",squash"`
+		pvMaxACPower       `mapstructure:",squash"`
+		batteryPowerLimits `mapstructure:",squash"`
```
```suggestion
		batteryCapacity    `mapstructure:",squash"`
		batteryPowerLimits `mapstructure:",squash"`
		pvMaxACPower       `mapstructure:",squash"`
```

andig commented on meter/lgess.go:
```diff
@@ -56,11 +57,11 @@ func NewLgEssFromConfig(other map[string]interface{}, essType lgpcs.Model) (api.
 
 // NewLgEss creates an LgEss Meter
-func NewLgEss(uri, usage, registration, password string, cache time.Duration, capacity func() float64, batterySocLimits batterySocLimits, essType lgpcs.Model) (api.Meter, error) {
+func NewLgEss(uri, usage, registration, password string, cache time.Duration, capacity func() float64, batterySocLimits batterySocLimits, essType lgpcs.Model, batteryPowerLimits batteryPowerLimits) (api.Meter, error) {
```
```suggestion
func NewLgEss(uri, usage, registration, password string, cache time.Duration, capacity func() float64, batterySocLimits batterySocLimits, batteryPowerLimits batteryPowerLimits, essType lgpcs.Model) (api.Meter, error) {
```

andig commented on meter/rct.go:
```diff
@@ -61,11 +62,11 @@ func NewRCTFromConfig(ctx context.Context, other map[string]interface{}) (api.Me
 	}
 
-	return NewRCT(ctx, cc.Uri, cc.Usage, cc.batterySocLimits, cc.MaxChargePower, cc.Cache, cc.ExternalPower, cc.batteryCapacity.Decorator())
+	return NewRCT(ctx, cc.Uri, cc.Usage, cc.batterySocLimits, cc.MaxChargePower, cc.Cache, cc.ExternalPower, cc.batteryCapacity.Decorator(), cc.batteryPowerLimits)
```
parameter nach vorne

Hofyyy replied:
Wo soll er denn hin? Auch nach Alphabet?

andig commented on templates/definition/meter/huawei-sun2000.yaml:
```diff
@@ -27,12 +27,12 @@ params:
   - name: capacity
     advanced: true
   - name: maxacpower
+  - name: maxchargepower
```
Änderung notwendig?

Hofyyy replied:
Ist eine Verschiebung wegen Alphabetic order.

andig commented on templates/definition/meter/marstek-venus.yaml:
```diff
@@ -30,10 +30,10 @@ params:
     help:
       en: 2500W is the maximum for Venus.
       de: 2500W ist das Maximum für Venus.
+  - name: maxdischargepower
```
Bitte die Zeilen drüber raus. Den Hilfstext brauchts nicht.

Hofyyy replied:
Ok hab einen weiteren genau so geändert, wo es nur um einen statischen wert aus einem Datenblatt ging. Mein Verständnis ist du willst hier nicht die Datenblätter der Produkte doppeln.

andig commented on templates/definition/meter/rct-power.yaml:
```diff
@@ -22,10 +22,11 @@ params:
     type: int
     default: 97
     advanced: true
+  - name: maxchargepower
```
alle nach cache

andig commented:
@premultiply mit diesem PR bekommen wir jetzt immer das Problem, dass das Limit für die Zwangsladung:

    default: 4200

Jetzt auch als Ladelimit für den Optimizer angenommen wird.

Hofyyy committed:
Update meter/e3dc.go

sort variables in alphabetic order.

Co-authored-by: andig <cpuidle@gmail.com>

Hofyyy committed:
Removed static help information from datasheets

Removed datasheet values from help.

Hofyyy committed:
Merge branch 'battery_charge_power' of https://github.com/Hofyyy/evcc into battery_charge_power

Hofyyy committed:
Moved maxcharge / discharge behind cache

- Alphabetic order

Hofyyy committed:
Changed order.

Hofyyy commented on meter/lgess.go:
```diff
@@ -80,17 +81,19 @@ func NewLgEss(uri, usage, registration, password string, cache time.Duration, ca
 	// decorate battery
 	var batterySoc func() (float64, error)
 	var batterySocLimiter func() (float64, float64)
+	var batteryPowerLimiter func() (float64, float64)
```
Den verstehe ich nicht. Soll es nur von 84 auf 82 nach vorne gezogen werden oder warum schmeisst du batteryPowerLimiter ganz raus?

premultiply commented:
Ja, das war uns ja auch schon klar dass das nur bedingt zusammen passt, wir aber hier erstmal einen Start haben wollten.

Denn hier muss eigentlich unterschieden werden was der Speicher maximal laden **kann** (Optimizer) und was er bei Zwangsladung maximal (aus dem Netz) laden **darf**. Bei Entladung und auch bei den SoC-Schwellen ist das leider ähnlich.

maatinh commented:
Beim goodwe-hybrid (und sicherlich einigen anderen) hängt die maximale Lade/Entladeleistung der Batterie davon ab wieviele Module übereinander gestapelt wurden und wie hoch dementsprechend die Batteriespannung ist. 

Ich lade meine Batterie an einem goodwe-hybrid aktuell netzdienlich und batterieschonend indem ich über modbus den maximalen Strom begrenze. Eine Leistung lässt sich nicht direkt einstellen. Das sollte bei der Implementierung eventuell berücksichtigt werden.

top-gun commented:
Wie ist es denn mit der benutzerdefinierten Batterie samt benutzerdefiniertem Wechselrichter?
